src/modules/recipe_preferences/RecipePreferencesProvider.ts

```diff
@@ -0,0 +1,86 @@
+import { Inject, Service } from "typedi";
+import { DATABASE, IDatabase } from "../../Database";
+import { User } from "../../model/User";
+import { RecipePreferences } from "../../model/RecipePreferences";
+import { Record } from "neo4j-driver";
+
+export const RECIPE_PREFERENCES_PROVIDER = "recipe-preferences-provider";
+
+export interface IRecipePreferencesProvider {
+  getRecipePreferences(user: User): Promise<RecipePreferences>;
+  setRecipePreferences(
+    pref: RecipePreferences,
+    user: User
+  ): Promise<RecipePreferences>;
+}
+
+@Service(RECIPE_PREFERENCES_PROVIDER)
+export class RecipePreferencesProvider implements IRecipePreferencesProvider {
+  constructor(@Inject(DATABASE) private readonly db: IDatabase) {}
+
+  async getRecipePreferences(user: User): Promise<RecipePreferences> {
+    const session = this.db.getSession();
+    try {
+      const result = await session.run(
+        "MATCH (user:User { uuid: $userId }) \n" +
+          "MATCH (user)-[:HAS_PREF]->(prefs) RETURN prefs",
+        {
+          userId: user.uuid,
+        }
+      );
+      return this._createRecipePrefsFromRecord(result.records, user);
+    } catch (e) {
+      return Promise.reject(e);
+    } finally {
+      await session.close();
+    }
+  }
+
+  async setRecipePreferences(
+    pref: RecipePreferences,
+    user: User
+  ): Promise<RecipePreferences> {
+    const session = this.db.getSession();
+    try {
+      const result = await session.run(
+        "MATCH (user:User { uuid: $userId }) \n" +
+          "OPTIONAL MATCH (user)-[r:HAS_PREF]->() \n" +
+          "MERGE (prefs:RecipePrefs {dairy: $dairy, gluten: $gluten, vegetarian: $vegetarian, vegan: $vegan}) \n" +
+          "CREATE (user)-[:HAS_PREF]->(prefs) \n" +
+          "DELETE r \n" +
+          "RETURN prefs",
```

Comment:
```suggestion
        "MATCH (user:User { uuid: $userId }) " + 
        "MERGE (user)-[r:HAS_PREF]->(prefs:RecipePrefs) " + 
        "SET prefs.dairy = $dairy " + 
        "SET prefs.gluten = $gluten " + 
        "SET prefs.vegetarian = $vegetarian " + 
        "SET prefs.vegan = $vegan " + 
        "RETURN prefs",
```
Bin mir zwar nicht ganz sicher, wie genau MERGE funktioniert, weil ich mich mit Cypher noch nicht so gut auskenne, aber beim Testen hat das bei mir dann immer ne neue Prefs-Node erstellt und bei der alten einfach nur die Relationship gelöscht, aber die alte Node selbst da gelassen. Deswegen würde ich eher das hier vorschlagen :)

Reply:
`MERGE` erstellt eine Node nur wenn sie noch nicht existiert. 
Dadurch das wir eine kleine begrenzte Anzahl an Einstellungsmöglichkeiten haben, kann ich die Nodes so lassen. Die die query werden die User Nodes immer nur mit den Passenden Preferences Nodes verknüpft. Sprich mehrere User können zeigen auf die gleiche Preferences Node.
Wenn sich die Preferences für einen User ändern, Lösche ich einfach die Relation und füge eine Relation zu einer anderen Node hinzu.